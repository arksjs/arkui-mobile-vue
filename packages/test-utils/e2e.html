<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
