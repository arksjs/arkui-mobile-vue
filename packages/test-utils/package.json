{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
