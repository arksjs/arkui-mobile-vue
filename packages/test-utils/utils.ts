export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
