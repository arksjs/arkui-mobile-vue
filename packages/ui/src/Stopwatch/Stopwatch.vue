<template>
  <div class="fx-stopwatch">
    <slot
      :time="countTime.time"
      :days="countTime.days"
      :fullHours="countTime.fullHours"
      :thousandsFullHours="countTime.thousandsFullHours"
      :hours="countTime.hours"
      :minutes="countTime.minutes"
      :seconds="countTime.seconds"
      :milliseconds="countTime.milliseconds"
    >
      {{
        parseInt(countTime.fullHours) > 0
          ? (thousands ? countTime.thousandsFullHours : countTime.fullHours) +
            ':'
          : ''
      }}{{ countTime.minutes }}:{{ countTime.seconds
      }}{{ showMilliseconds ? '.' + countTime.milliseconds : '' }}
    </slot>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import type { CountTime } from '../CountDown/types'
import { getCountTime } from '../CountDown/util'
import { useCountTime } from '../CountDown/use-count-time'
import { cloneData, returnTrue } from '../helpers/util'
import type { PropsToEmits } from '../helpers/types'
import type { StopwatchEmits } from './types'

export default defineComponent({
  name: 'fx-stopwatch',
  props: {
    // 是否显示毫秒数
    showMilliseconds: {
      type: Boolean,
      default: true
    },
    // 是否千分位显示
    thousands: {
      type: Boolean,
      default: true
    }
  },
  emits: {
    start: returnTrue,
    reset: returnTrue,
    stop: payload => payload && payload.detail && Array.isArray(payload.laps)
  } as PropsToEmits<StopwatchEmits>,
  setup(props, { emit }) {
    let time = 0
    let startTime = 0
    let laps: CountTime[] = []

    const { times, timeStop, timeUpdate, timeStart } = useCountTime(
      ({ update }) => {
        time = Date.now() - startTime
        update(time)
      }
    )

    function doing() {
      return startTime > 0
    }

    function stop() {
      if (doing()) {
        const _laps = cloneData(laps)
        _laps.push(getNextLap())

        emit('stop', {
          detail: getCountTime(time),
          laps: _laps
        })

        startTime = 0
        timeStop()
      }
    }

    function start() {
      if (startTime === 0) {
        startTime = Date.now() - time
      }

      emit('start')

      timeStart()
    }

    function reset() {
      doing() && stop()

      emit('reset')

      time = 0
      laps = []
      timeUpdate(0)

      timeStop()
    }

    function getNextLap() {
      let totalTime = 0

      laps.forEach(({ time }) => {
        totalTime += time
      })

      return getCountTime(time - totalTime)
    }

    function lap(): CountTime[] {
      if (doing()) {
        laps.push(getNextLap())

        return cloneData(laps)
      }

      return []
    }

    timeUpdate(0)

    return {
      time,
      countTime: times,
      stop,
      start,
      reset,
      lap
    }
  }
})
</script>
