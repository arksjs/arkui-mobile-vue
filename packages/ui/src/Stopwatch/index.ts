import { withInstall } from '../helpers/with-install'
import Stopwatch from './Stopwatch.vue'

export { Stopwatch }
export const install = withInstall(Stopwatch)
export default Stopwatch
