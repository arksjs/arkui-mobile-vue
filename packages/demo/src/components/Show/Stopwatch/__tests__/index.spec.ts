import { shallowMount } from '@vue/test-utils'
import { Stopwatch } from '@/Stopwatch'

describe('Stopwatch', () => {
  test('should render default correctly', () => {
    const wrapper = shallowMount(Stopwatch, {})

    expect(wrapper.html()).toMatchSnapshot()
  })
})
