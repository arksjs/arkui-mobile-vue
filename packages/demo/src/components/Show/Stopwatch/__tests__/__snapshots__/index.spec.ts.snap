// Vitest Snapshot v1

exports[`Stopwatch > should render default correctly 1`] = `"<div class=\\"fx-stopwatch\\">00:00.000</div>"`;
